### ISSUE TITLE
CA2213: Interlocked Exchange then Dispose

### ISSUE THREAD

**Issue by skarllot** [dotnet/roslyn-analyzers#2892]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.5

#### Diagnostic ID

[CA2213](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2213-disposable-fields-should-be-disposed?view=vs-2019)

#### Repro steps

Create the following code:
```csharp
public sealed class TestClass : IDisposable
{
    private ManualResetEventSlim readyEvent;
    public void Dispose() => Interlocked.Exchange(ref readyEvent, null)?.Dispose();
}
```

#### Expected behavior

Recognize thread safe calls.

#### Actual behavior

Does not detect that field is being disposed safely.

**Comment by brian-reichle** [comment]
Maybe it's just a bad example, but using using `Interlocked` and `Dispose` like that seems very wrong to me. The caller should not be disposing an object that is in use by another thread (callers responsibility to ensure this) and if no other thread is using the object when Dispose is called, then the use of `Interlocked` seems redundant.

**Comment by goyzhang** [comment]
I think current CA2213 analyzer only do semantical check that if you call `Dispose()` directly on that field.
It does not have any knowledge about the return value of `Interlocked.Exchange`(and it shouldn't care either). The return value is just a `ManualResetEventSlim` type instance for the analyzer, it doesn't necessary to be the same reference and the analyzer prefer making no assumption about the implementation details of `Interlocked.Exchange`(although that in fact it does return the original value of readyEvent).

**Comment by skarllot** [comment]
I agree with @brian-reichle, should I close this issue?
